core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/dto/NameOrCrn.java

```diff
@@ -0,0 +1,87 @@
+package com.sequenceiq.cloudbreak.api.endpoint.v4.dto;
+
+import static com.sequenceiq.cloudbreak.util.NullUtil.throwIfNull;
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class NameOrCrn {
+
+    private static final String NULL_DTO_EXCEPTION_MESSAGE = "Name or crn should not be null.";
+
+    private static final String NAME_MUST_BE_PROVIDED_EXCEPTION_MESSAGE = "Name must be provided.";
+
+    private static final String CRN_MUST_BE_PROVIDED_EXCEPTION_MESSAGE = "Crn must be provided.";
+
+    @VisibleForTesting
+    final String name;
+
+    @VisibleForTesting
+    final String crn;
+
+    private NameOrCrn(String name, String crn) {
+        this.name = name;
+        this.crn = crn;
+    }
+
+    public static NameOrCrn ofName(String name) {
+        if (StringUtils.isEmpty(name)) {
+            throw new IllegalArgumentException(NAME_MUST_BE_PROVIDED_EXCEPTION_MESSAGE);
+        }
+        return new NameOrCrn(name, null);
+    }
+
+    public static NameOrCrn ofCrn(String crn) {
+        if (StringUtils.isEmpty(crn)) {
+            throw new IllegalArgumentException(CRN_MUST_BE_PROVIDED_EXCEPTION_MESSAGE);
+        }
+        return new NameOrCrn(null, crn);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder toString = new StringBuilder();
+        toString.append("[NameOrCrn");
+        if (isNotEmpty(name)) {
+            toString.append(" of name: '");
+            toString.append(name);
+        } else {
+            toString.append(" of crn: '");
+            toString.append(crn);
+        }
+        toString.append("']");
+        return toString.toString();
+    }
+
+    public static class NameOrCrnReader {
+
+        private final NameOrCrn nameOrCrn;
+
+        private NameOrCrnReader(NameOrCrn nameOrCrn) {
+            this.nameOrCrn = nameOrCrn;
+        }
+
+        public static NameOrCrnReader create(NameOrCrn nameOrCrn) {
+            throwIfNull(nameOrCrn, () -> new IllegalArgumentException(NULL_DTO_EXCEPTION_MESSAGE));
+            return new NameOrCrnReader(nameOrCrn);
+        }
+
+        public boolean hasName() {
```

Comment:
Why not moving the methods if this inner class to the original class, and remove `NameOrCrnReader`? Why would we bother instantiating this to be able to read the `NameOrCrn` class?

Reply:
The sole purpose of this inner-class thing is an enforced null check. The original code did check for nulls. I did think a while whether to do just delete the null check (and so risk an NPE), or keep it. The (small) added value I see in enforcing null check is the opportunity to better log the error and so help diagnostics.

Reply:
I see. I still have doubts regarding whether it will provide us any real life diagnostic assistance. I think in most cases without this utility we would be able to identify the source of the NPE right away.

Reply:
Yeah, I have doubts as well, we do not win much. Ok, so if @gregito agrees, I would then remove the null checks. Please confirm.

Reply:
IMHO NPE should never occur. if the only purpose of this NameOrCrnReader class is to ensure that nobody can create an invalid NameOrCrn instance then the null check should be there. on the other hand, null and emptiness checks should be performed at the same time (StringUtils.isEmpty also checks for null), hence one of the checks can be abandoned.

Reply:
@gregito You cannot create an empty NameOrCrn object, it either contains the name or the crn. If name is null when you use NameOrCrn.ofName(name), then you get an exception right away (and the same for crn). The only thing NameOrCrn cannot guard for is that instead of creating it you pass null (could happen accidentally).
Another guard the class does not contain but could: if NameOrCrn was create with name, then reading the crn should throw. I think that would make sense.

Reply:
Change done.